text/0000-glimmer-component-signature.md

```diff
@@ -0,0 +1,361 @@
+
+For all three signature members proposed here, Glint will align its API with whatever terminology this RFC lands on.
+
+### Migration
+
+We can implement this change in `@glimmer/component` 1.x in a backwards compatible way, allowing for a deprecation period before moving exclusively to the signature approach in 2.0. Because capitalized arg names are currently illegal, any valid signature will not represent valid arg names, so the backing class can [accept both formats](https://tsplay.dev/Nr2rzN) and distinguish which was intended.
+
+Ideally we'll be able to encourage authors to migrate their usage of `Args` in the same ways they're used to being nudged to move away from other deprecated APIs in the Ember ecosystem.
+
+#### Codemods
+
+A simple codemod to turn `Component<MyArgs>` into `Component<{ Args: MyArgs }>` would be straightforward, but it may also be feasible to migrate more completely by inferring information like block names and where `...splattributes` are used from colocated templates. The richness we pursue here will likely depend on the appetite of the community to explore what's possible.
+
+#### Deprecation Warnings/Linting
+
+While type-only deprecations aren't something the Ember ecosystem has dealt with much previously, it's a muscle we may want to begin building as [official TypeScript support] is under consideration.
+
+The `@typescript-eslint` suite of packages supports writing [type-aware rules] in projects that provide appropriate configuration in their `eslintrc`, so one option available to use would be to use that functionality to allow users to lint for components that haven't yet been migrated to use a signature type. We could expose this as part of a standalone ESLint plugin, or perhaps make it available on an opt-in basis in `eslint-plugin-ember`.
+
+[official TypeScript support]: https://github.com/emberjs/rfcs/pull/724
+[type-aware rules]: https://github.com/typescript-eslint/typescript-eslint#can-we-write-rules-which-leverage-type-information
+
+## Drawbacks
+
+As with any change that deprecates supported behavior, there's an inherent cost associated with migrating the user base over to new patterns. One goal of this change, however, is to ease such migrations in the future: as the templating system evolves and new information becomes relevant to capture and document, the `Signature` type provides a place for that information to live without disrupting existing code.
+
+The other potential drawback to this approach is that it introduces TypeScript type information that has no visible effect on the component using out-of-the-box tooling. Without a template-aware system like Glint or `els-addon-typed-templates` for validating components, nothing enforces that the signature declared is actually accurate. See the "Only `Args`" section under Alternatives below for further discussion of this point.
+
+## Alternatives
```

Comment:
Worth copying over alternative, non-typescript ideas from the original RFC? Like json-schema for this?

Reply:
I don't think so. This RFC is very intentionally _not_ aiming to answer the broad "what is the be-all end-all way to document components" question you lay out in https://github.com/emberjs/rfcs/pull/678. It's proposing a narrow, focused change to a specific API, and the potential usefulness for documentation is one of several motivating use cases for the change.

Similarly, you raised applicability to template-only components and `.js` files in Discord as something that's missing here, but that's something that's intentionally outside the scope of the change being proposed here. Changing `Args` to `Signature` doesn't change anything about the feasibility of using JSDoc or `.d.ts` files to capture type information for non-`.ts` components. That question is the same with or without this RFC; the only thing that changes is the volume of information we're encoding.

All of that is to say that I don't think this RFC can (or is attempting to) supercede #678. You're attempting to tackle a much broader set of topics there, and while I think the changes here will affect that RFC if they're accepted, merging this RFC wouldn't answer all the questions posed in that one 🙂